src/core/storage/fileio/s3_api.cpp

```diff
@@ -346,249 +397,268 @@ list_objects_response list_objects_impl(s3url parsed_url,
+            ret.objects_last_modified.push_back(stream.str());
+            ret.objects_size.push_back(o.GetSize());
+          }
+
+          // now iterate through common prefixes - these are directories
+          Aws::Vector<Aws::S3::Model::CommonPrefix> prefixes;
+          prefixes = result.GetCommonPrefixes();
+          for (auto const& p : prefixes) {
+            std::string key = std::string(p.GetPrefix().c_str());
+            // strip the ending "/" on a directory
+            if (boost::ends_with(key, "/"))
+              key = key.substr(0, key.length() - 1);
+            ret.directories.push_back(key);
+          }
+
+          // more results to retrieve
+          moreResults = result.GetIsTruncated();
+          if (moreResults) {
+            // add to the request object with continuation token
+            request.WithContinuationToken(result.GetContinuationToken());
+          }
+
+          // jump out the retry loop
+          break;
+
+        } else {
+          auto error = outcome.GetError().GetResponseCode();
+
+          if (error == Aws::Http::HttpResponseCode::TOO_MANY_REQUESTS) {
+            n_retry++;
+            std::this_thread::sleep_for(std::chrono::milliseconds(backoff));
+            backoff *= 2;
+          } else {
             std::stringstream ss;
             reportS3ErrorDetailed(ss, parsed_url, S3Operation::List,
                                   clientConfiguration, outcome)
                 << std::endl;
             ret.error = ss.str();
+            logstream(LOG_DEBUG)
+                << "list_objects_impl failed:" << ret.error << std::endl;
+            // no need to retry, just jump out all while loop
+            break;
+          }
         }
+      } while (n_retry < 3);  // finished retry
 
-    } while (moreResults);
+    } while (moreResults && success);
 
-    Aws::ShutdownAPI(options);
     for (auto& dir : ret.directories) {
       s3url dirurl = parsed_url;
       dirurl.object_name = dir;
-      dir = string_from_s3url(dirurl);
+      // this is not necessary to override what returned by s3
+      dir = dirurl.string_from_s3url();
     }
-    for (auto& object: ret.objects) {
+    for (auto& object : ret.objects) {
       s3url objurl = parsed_url;
       objurl.object_name = object;
-      object = string_from_s3url(objurl);
+      object = objurl.string_from_s3url();
     }
     return ret;
 }
 
 /// returns an error string on failure. Empty string on success
-std::string delete_object_impl(s3url parsed_url,
-                               std::string proxy,
+std::string delete_object_impl(s3url parsed_url, std::string proxy,
                                std::string endpoint) {
-    std::string ret;
+  std::string ret;
 
-    // initialization
-    Aws::SDKOptions options;
-    options.loggingOptions.logLevel = Aws::Utils::Logging::LogLevel::Info;
-    Aws::InitAPI(options);
+  // credentials
+  Aws::Auth::AWSCredentials credentials(parsed_url.access_key_id.c_str(),
```

Comment:
Not a comment on you, but it seems like there could be a lot of logic in some of these places consolidated into common logic functions.  Lots of duplicated config code.

Reply:
Yeah, it's one of my complaints. Just leave it here for people to fix it in the future.

Reply:
Why later and not now?   It's hard to review copy-pasted code.

Reply:
due to the time constraint, I don't want to break things that are working now. The logic was there for a long time and we can push to the next release to clean up code.